### PR TITLE
Refactor aggregation 

### PR DESCRIPTION
**Pull request by MRossol**
Create abstract and hierarchical SCPoint and Aggregation classes
Add a general Aggregation class (for Power Rose aggregation)

### CODE REVIEW

**Comment by MRossol** [comment]
@grantbuster  Please take a look at the points.py and point_summary.py hierarchy.

File: reV/supply_curve/points.py [hunk 1]
```diff
@@ -43,109 +1043,71 @@ def __init__(self, gid, excl, gen, tm_dset, gen_index, excl_dict=None,
...
+        self._gen_fpath, self._gen = self._h5_fpath, self._h5
         self._gen_gids, self._res_gids = self._parse_techmap(tm_dset,
                                                              gen_index)
         self._remove_offshore(self._gen_gids, self._res_gids,
                               offshore_flags=offshore_flags)
 
-    def _parse_files(self, excl, gen):
-        """Parse gen + excl filepath input or handler object and set to attrs.
+    @staticmethod
+    def _parse_h5_file(gen):
```
Comment (grantbuster) [inline reV/supply_curve/points.py]: This method is redundant to the static method in `AggregationSupplyCurvePoint`

File: reV/supply_curve/points.py [hunk 2]
```diff
@@ -2,27 +2,1027 @@
...
+        kwargs = {"excl_dict": excl_dict, "resolution": resolution,
+                  "exclusion_shape": exclusion_shape, "close": close}
+        with cls(gid, excl, tm_dset, **kwargs) as point:
+            agg = point.aggregate(data)
+
+        return agg
+
+
+class AggregationSupplyCurvePoint(SupplyCurvePoint):
+    """Generic single SC point based on exclusions, resolution, and techmap"""
```
Comment (grantbuster) [inline reV/supply_curve/points.py]: Update docstring: `"""Generic single SC point to aggregate data from an h5 file."""`

File: reV/supply_curve/points.py [hunk 3]
```diff
@@ -2,27 +2,1027 @@
...
+        else:
+            raise SupplyCurveInputError('SupplyCurvePoints needs a '
+                                        '.h5 file path, or '
+                                        'Resource handler but '
+                                        'received: {}'
+                                        .format(type(h5)))
+
+        return h5_fpath, h5
+
+    def __enter__(self):
```
Comment (grantbuster) [inline reV/supply_curve/points.py]: I think these inherited (redundant)

File: reV/supply_curve/points.py [hunk 4]
```diff
@@ -2,27 +2,1027 @@
...
+        resolution : int
+            Number of exclusion points per SC point along an axis.
+            This number**2 is the total number of exclusion points per
+            SC point.
+        excl_area : float
+            Area of an exclusion cell (square km).
+        exclusion_shape : tuple
+            Shape of the full exclusions extent (rows, cols). Inputing this
+            will speed things up considerably.
+        offshore_flags : np.ndarray | None
```
Comment (grantbuster) [inline reV/supply_curve/points.py]: remove offshore_flags from docstring

File: reV/supply_curve/aggregation.py [hunk 5]
```diff
@@ -1,64 +1,39 @@
...
-                                      OutputWarning, FileInputError,
-                                      InputWarning, SupplyCurveInputError)
+                                      FileInputError)
 
 
 logger = logging.getLogger(__name__)
 
 
-class AggFileHandler:
+class AbstractAggFileHandler(ABC):
```
Comment (grantbuster) [inline reV/supply_curve/aggregation.py]: The abstract vs. concrete versions of this seem redundant. Maybe differentiate the docstrings to specify use cases?

File: reV/supply_curve/aggregation.py [hunk 6]
```diff
@@ -354,228 +194,58 @@ def _check_files(self):
...
-    def _serial_summary(excl_fpath, gen_fpath, tm_dset, gen_index,
-                        excl_dict=None, res_class_dset=None,
-                        res_class_bins=None, cf_dset='cf_mean-means',
-                        lcoe_dset='lcoe_fcr-means', data_layers=None,
-                        resolution=64, power_density=None, friction_fpath=None,
-                        friction_dset=None, gids=None,
-                        area_filter_kernel='queen', min_area=None, args=None,
-                        ex_area=0.0081, close=False, check_excl_layers=False):
+    @abstractmethod
+    def run_serial(sc_point_method, excl_fpath, tm_dset,
```
Comment (grantbuster) [inline reV/supply_curve/aggregation.py]: I learned there's actually a `abstractstaticmethod` so you don't have to stack abtractmethod and staticmethod

File: reV/supply_curve/sc_aggregation.py [hunk 7]
```diff
@@ -0,0 +1,1057 @@
...
+from reV.utilities.execution import SpawnProcessPool
+from reV.utilities.exceptions import (EmptySupplyCurvePointError,
+                                      OutputWarning, FileInputError,
+                                      InputWarning, SupplyCurveInputError)
+
+logger = logging.getLogger(__name__)
+
+
+class SupplyCurveAggFileHandler(AbstractAggFileHandler):
+    """Simple framework to handle aggregation file context managers."""
```
Comment (grantbuster) [inline reV/supply_curve/sc_aggregation.py]: Modify doc string to differentiate from abstract version.

File: reV/supply_curve/points.py (hunk 2 above)
Comment (MRossol) [reply reV/supply_curve/points.py]: good catch

File: reV/supply_curve/points.py (hunk 3 above)
Comment (MRossol) [reply reV/supply_curve/points.py]: removed

File: reV/supply_curve/points.py (hunk 4 above)
Comment (MRossol) [reply reV/supply_curve/points.py]: whoops

File: reV/supply_curve/points.py (hunk 1 above)
Comment (MRossol) [reply reV/supply_curve/points.py]: Not quite, in AggregationSupplyCurvePoint we initialize a Resource handler, while in GenSCPoint we initialize an Outputs handler.  I honestly not sure if we need the Outputs handler so we could remove this...

File: reV/supply_curve/aggregation.py (hunk 5 above)
Comment (MRossol) [reply reV/supply_curve/aggregation.py]: the main difference is we added the (Resource) h5 file handler

File: reV/supply_curve/aggregation.py (hunk 6 above)
Comment (MRossol) [reply reV/supply_curve/aggregation.py]: cool!

File: reV/supply_curve/sc_aggregation.py (hunk 7 above)
Comment (MRossol) [reply reV/supply_curve/sc_aggregation.py]: done

File: reV/supply_curve/points.py (hunk 1 above)
Comment (grantbuster) [reply reV/supply_curve/points.py]: Yeah I dont think we need the outputs handler but its not a big deal either way

**Pull request merged**